pkg/controller/usersignup/usersignup_controller.go

```diff
@@ -497,9 +518,41 @@ func (r *ReconcileUserSignup) provisionMasterUserRecord(userSignup *toolchainv1a
 		return r.wrapErrorWithStatusUpdate(logger, userSignup, r.setStatusFailedToCreateMUR, err,
 			"Error creating MasterUserRecord")
 	}
+	logger.Info("Created MasterUserRecord", "Name", mur.Name, "TargetCluster", targetCluster)
+	// increment the counter of MasterUserRecords
 	counter.IncrementMasterUserRecordCount()
+	// also, update the 'toolchain.dev.openshift.com/activation-counter' annotation value
+	return r.ensureActivationCounter(userSignup, logger)
+}
 
-	logger.Info("Created MasterUserRecord", "Name", mur.Name, "TargetCluster", targetCluster)
+// ensureActivationCounter increments the 'toolchain.dev.openshift.com/activation-counter' annotation value on the given UserSignup
+func (r *ReconcileUserSignup) ensureActivationCounter(userSignup *toolchainv1alpha1.UserSignup, logger logr.Logger) error {
+	// increment the counter of Returning Users
+	c := 1
+	if activations, exists := userSignup.Annotations[toolchainv1alpha1.UserSignupActivationCounterAnnotationKey]; exists {
+		logger.Info("updating 'toolchain.dev.openshift.com/activation-counter' on active user")
+		if activations, err := strconv.Atoi(activations); err == nil {
+			// increment the value of the annotation
+			activations++
+			userSignup.Annotations[toolchainv1alpha1.UserSignupActivationCounterAnnotationKey] = strconv.Itoa(activations)
+			c = activations
+		} else {
+			logger.Error(err, "The 'toolchain.dev.openshift.com/activation-counter' annotation value was not an integer and was reset to '1'.", "value", activations)
+			// "best effort": reset number of activations to 1 for this user
+			userSignup.Annotations[toolchainv1alpha1.UserSignupActivationCounterAnnotationKey] = "1"
+		}
+	} else {
+		// annotation was missing so assume it's the first activation
+		logger.Info("setting 'toolchain.dev.openshift.com/activation-counter' on new active user")
+		userSignup.Annotations[toolchainv1alpha1.UserSignupActivationCounterAnnotationKey] = "1" // first activation, annotation did not exist
+	}
+	// will not trigger a reconcile if update succeeds (see UserSignupChangedPredicate)
+	err := r.client.Update(context.TODO(), userSignup)
+	if err != nil {
+		return err
+	}
```

Comment:
done in 48c4932

Reply:
Thanks for moving the code, but this still doesn't solve the main problem - if the update of the UserSignup with the modified annotations fails, then you never increment the counter.
As I wrote above - we should update the annotation together with the label - for example something like this:
```go
func (r *ReconcileUserSignup) setStateLabel(reqLogger logr.Logger, userSignup *toolchainv1alpha1.UserSignup, value string) error {
	oldValue := userSignup.Labels[toolchainv1alpha1.UserSignupStateLabelKey]
	if oldValue == value {
		// skipping
		return nil
	}
	userSignup.Labels[toolchainv1alpha1.UserSignupStateLabelKey] = value
	if value == toolchainv1alpha1.UserSignupStateLabelValueApproved {
		if err := r.ensureActivationCounter(reqLogger, userSignup); err != nil {
			return err
		}
	} else {
		if err := r.client.Update(context.TODO(), userSignup); err != nil {
			return r.wrapErrorWithStatusUpdate(reqLogger, userSignup, r.setStatusFailedToUpdateStateLabel, err,
				"unable to update state label at UserSignup resource")
		}
	}
	updateMetricsByState(oldValue, value)

	return nil
}
```

Reply:
you can modify the code to make it more clear of course - but I hope that you got my point.

Reply:
I don't understand your solution:

1. `if value == toolchainv1alpha1.UserSignupStateLabelValueApproved { ... }` will update the activation counter but it will skip the usersignup update?
2. if the usersignup update fails, the call to `updateMetricsByState` will be skipped, but the `ensureActivationCounter` should be done nonetheless?

I thought that in https://github.com/xcoulon/host-operator/blob/IssueCRT988_metric_user_reactivations/pkg/controller/usersignup/usersignup_controller.go#L314-L317, we would ensure that the state label is correctly set (even if it fails before)

Reply:
ad 1. not really - we update the label before the if statement:
```
	userSignup.Labels[toolchainv1alpha1.UserSignupStateLabelKey] = value
```
so in the `ensureActivationCounter` when we update the UserSignup, we update also the annotation.

ad 2. if the update fails, then nothing will be incremented nor updated - that is solved in the `ensureActivationCounter` logic, right?

> I thought that in https://github.com/xcoulon/host-operator/blob/IssueCRT988_metric_user_reactivations/pkg/controller/usersignup/usersignup_controller.go#L314-L317, we would ensure that the state label is correctly set (even if it fails before)

yes, but that doesn't apply to update of the annotation. In your original code - if the update of the annotation fail inside `ensureActivationCounter` then the next reconcile won't try to update it again because the label is already updated to the latest version. 
We need to make the update of both label and annotation atomic - both things needs to be updated at once.

I guess that the best way to verify it is to create a test case that will fail for the first attempt to update the annotation and then execute the reconcile again and see if the annotation will be updated or not.

Reply:
> We need to make the update of both label and annotation atomic - both things needs to be updated at once.

yes, I know, but I think that the code is getting a bit confusing :(

Reply:
refactored in 9960c92, hopefully this is more explicit and the updated test will show that the metric is not updated in case of failure

Reply:
> yes, I know, but I think that the code is getting a bit confusing because the ensureActivationCounter method not only sets/updates the annotation value on the UserSignup, but it also updates the UserSignup resource on the cluster. I will refactor that to improve readability

I know - this is why I said:

> you can modify the code to make it more clear of course - but I hope that you got my point.

The code that I wrote in the comment was just an example to show the goal of the change - it wasn't supposed to be the final solution at all ;-) 

> refactored in 9960c92, hopefully this is more explicit and the updated test will show that the metric is not updated in case of failure

cool, I'll take a look at it on Monday :+1: